doc/evaluation_design.md

```diff
@@ -0,0 +1,64 @@
+
+![evaluate_flowchart](/doc/figures/train_and_evaluate_flowchart.svg)
+<em>Figure 2 </em>. Flowchart of Train and Evaluation in Worker
+
+While the worker is performing train step N, the Dataset.prefetch thread is preparing the data for step N+1. If step N is the last batch in the task, the generator in the worker for dataset will get next task until the task mode is train. Then worker generates a train batch for step N+1 by Dataset.prefetch. 
+
+There are two problems in the existing evaluation design:
+
+* Problem 1: One worker will get all evaluation tasks in the flowchart shown as <em>Figure 2</em>.\
+As shown in the 10th step of <em> Figure 1.</em>, the master generates evaluation tasks and insert them into the head of todo queue when an evaluation jon starts. When one worker starts to process the last mini-batch in its task, it will get next task until the task mode is train to prepare next mini-batch for Dataset.prefetch. The result is that the worker will get all evaluation tasks before it gets next train task. Meanwhile, the master will remove all evaluation tasks from todo to doing list and other worker can not get any evaluation task. So, there is only one work to serially process evaluation tasks.
+
+* Problem 2: 
+For current ElasticDL, workers may use different model versions for evaluation for embedding vectors. For non-embedding variables, workers will use the same version from checkpoint.\
+As shown in <em>Figure 1</em>. The worker 1 gets model from the pserver at 15th step and the worker 2 gets model later at 25th step for the reason that the worker 2 is slower. However, the pserver has updated the model by gradients from the worker 2 before the worker2 gets the model to evaluate. So records assigned to different worker in a same validation dataset will be evluated with different model verion. 
+
+
+## Evaluate with Current Model in Pserver 
+1. The pserver updates model variables if it receives gradient from the worker after an evaluation job starts.\
+The solution is the same as existing desigin in ElasticDL which can not resolve Problem 2. In order to resolve Problem 1, the master can divide train and evaluation tasks into two todo lists at 10th step in <em>Figure 1</em>. When evalution starts, the master generates evaluation tasks and inserts those into evluation todo list not train list. Meanwhile the master needs to tell all pservers to change the model mode to evaluation.  The worker starts to pull evaluation tasks and inferences outputs when the mode of model it gets is evluation. The pserver updates model variables by the mini-batch gradients from each worker at most once. Because the model mode is evluating when the worker get for next mini-batch. The master will change the model mode to train after receiving all record outputs of evluation tasks. Then worker(s) continues to process training tasks.
```

Comment:
May also need to consider situations where pservers might fail and the two task lists are lost/incomplete

Reply:
The task will fail because the worker can not get the model from ps. Then, the worker will retry the task like the training task.